### ISSUE TITLE
[Bug] Image upload prepends the full path to the image, when unchanged and editing a crud

### ISSUE THREAD

**Issue by PaulLawton227** [Laravel-Backpack/CRUD#1372]
# Bug report

I'm using image upload type in a crud field (with a mutator to upload the file instead of encoding base64 in the DB). When editing a crud - if the image has not changed - then the path to the image (stored in the DB - e.g. uploads/[filename]) is prepended with the full site url e.g. https://mysite.com/uploads/[filename] and saved in the DB as he full url. 

### What I did:

Add a crud for a model with a type of image e.g.
```php
$this->crud->addField([ // image
            'label' => "Map Image",
            'name' => "map_image",
            'type' => 'image',
            'upload' => true,
            'disk' => 's3',
]);
```

and on the modal - I have a mutator to upload the image and return the path to store in the database
e.g. 
```php
public function setMapImageAttribute($value)
	{
		$attribute_name = "map_image";
		$disk = "s3";
		$this->processUpload($value, $attribute_name, $disk);
	}
```

### What I expected to happen:
When editing with a crud - if an image has been previously uploaded (via CRUD add), then submitting the form - it should not modify the image path stored in the DB.  

### What happened:
The form is submitted and the value to store in the DB is taken from the image src - which is http://mysite.com/uploads/[filename] whereas it should just use the $value and store uploads/[filename] 

### What I've already tried to fix it:

In image.blade.php - I replaced:
$hiddenImage.val($mainImage.attr('src'));
with 
$hiddenImage.val('{{ $value }}'); // use the initial value instead of the full src.

This seems to have fixed the problem - but I've love to know if this is a correct fix and doesn't affect anything else.

### Backpack, Laravel, PHP, DB version:
backpack/crud version 3.4.6

**Comment by PaulLawton227** [comment]
Ok, after some more testing this doesn't work nicely will multiple image uploads.  Will investigate further and report back

**Comment by PaulLawton227** [comment]
I have a fix for multiple images. Ignore previous updates - all you need to do is the following:

In image.blade.php

Add a data element to:
`<img id="mainImage" src="{{ $image_url }}" data-value="{{ $value }}">`

and the line:
`$hiddenImage.val($mainImage.attr('src'));`
needs changing to:
`$hiddenImage.val($mainImage.data('value'));`

**Comment by dannoso** [comment]
@PaulLawton227 this is great why don't you share it with other users and submit a PR?

**Comment by androidedev** [comment]
I faced the exact same issue, but I take another path to solve it.

Being your solution a good one I think that it forces to load al base64 images in edit CRUD and that can be slow if there are many, and besides that I really against modify backpack if I can avoid it.

My aproach was to create a custom validator and use it in image uploads like that :

...
'image' => [
   'nullable',
   ....
   new BackPackImageFieldValidationRule(["image/png","image/jpeg"],'uploads')
],
...

The code for the validator is :

---------------
`<?php namespace App\CustomValidators;

      use finfo;
      use Illuminate\Support\Facades\Storage;
      use Illuminate\Contracts\Validation\Rule;
      use Intervention\Image;

      // Note :
      // You must implement "ImplicitRule" instead of "Rule" if you want that the validator 
      // be called when the value is null or not present, is not my case so I implement "Rule"
      class BackPackImageFieldValidationRule implements Rule
      {

          private $allowed;
          private $storagedisk;

          public function __construct($allowedmimes,$storagedisk) {
              $this->allowed = $allowedmimes;
              $this->storagedisk = $storagedisk;
          }

          // Here backpack comes with $value in base64 or a filename, comes with a filename if you 
          // edit and save in CRUD with already has an image and you don't touch it
          public function passes($attribute, $value)
          {
              if(!isset($value)) // this only takes effect if you implement "ImplicitRule"
                  return true;

              $allowedmime = false;

              // Check if $value is a base64 and has the allowed MIME
              foreach ($this->allowed as $mimevalue)
              {
                  $muststartwith = 'data:'.$mimevalue;
                  if( starts_with($value, $muststartwith) )
                  {
                      $allowedmime = true;
                      break;
                  }
              }

              if($allowedmime)
              {
                  // appears to be a valid base64 image, let's try to parse as an image
                  try {
                      Image\ImageManagerStatic::make($value);
                      return true;
                  }
                  catch (\Exception $e) {
                      return false;
                  }
              } else
              {
                  // is not a allowed mime so let's try other things ...
                  
                  // first check if is a base64 encoded if not is not a valid one
                  if(starts_with($value, 'data:'))
                  {
                      return false;
                  }
                  else
                  {
                      // if we are here is not a base64 encoded value this happens because BackPack
                      // image field always sends a base64 encoded image except when you edit a record 
                      // that already has an image in edit CRUD view and the view is saved without 
                      // touching the image, if that is the case the incoming value is the URL pointing 
                      // to the existing file. 
                      // In theory we can return true here because the image has not changed but is safer 
                      // to check al least
                      //  1- if the file name is really one that was stored in our file system
                      //  2- its mime type


                      // if we don't have any disk to do the check exit with false
                      if(!isset($this->storagedisk))
                          return false;

                      try {
                          // Extract url path
                          $imagepath = parse_url($value,PHP_URL_PATH);
                          // Remove "disk"
                          $imagepathwithoutdisk = substr($imagepath,strlen('/'.$this->storagedisk.'/'));
                          // We ask for existance, couse if not exist something is really wrong
                          if(!Storage::disk($this->storagedisk)->exists($imagepathwithoutdisk))
                              return false;
                          // Let's check the MIME 
                          $storagediskpath = public_path($this->storagedisk);
                          $hardpathtofile = $storagediskpath.'/'.$imagepathwithoutdisk;
                          $finfo = new finfo(FILEINFO_MIME_TYPE);
                          $mime = $finfo->file($hardpathtofile);
                          $res = in_array($mime,  $this->allowed, true);
                          if(!$res)
                              return false;
                          return true;
                      }
                      catch (\Exception $e) {
                          return false;
                      }
                  }
              }
          }

          public function message()
          {
              return 'The image is not a valid image, please check that is an allowed image type';
          }

      }`
---------------
      
      
Please note that this code is a bit paranoid, we can avoid the re-checking for the MIME time in second part and simply do :

`
// We ask for existance, couse if not exist something is really wrong
if(!Storage::disk($this->storagedisk)->exists($imagepathwithoutdisk))
   return false;
else
   return true;
`

Or simply return true for that case, after all we can assume that the file has not changed, so it must be in our filesystem, but I prefer to re-check.
                    

PS: I don't make a pull request because I don't see that backpack cover validators, and to be honest my backgroud is C#/Python I'm new in php and I'm not very confident with my php code :)

**Comment by lloy0076** [comment]
@androidedev Feel free to make an appropriate PR even if your PHP isn't perfect - someone will review it and it's easier to "fix" imperfect code than reinvent it from scratch.

Just be aware that most of the team are gentle reviewers; I can come across as a little persnickety and nit-picky but I try to be generally helpful.

**Comment by thomascharbit** [comment]
I have the same issue, i am wondering if this is a confirmed bug or an inappropriate use of the image field?

**Comment by tabacitu** [comment]
Wow - I'm so late to the party. Sorry guys, seems like this one fell through the cracks. I just tested it in Backpack 3.6 and it seems like we've solved it in the meantime:

> When editing with a crud - if an image has been previously uploaded (via CRUD add), then submitting the form - it should not modify the image path stored in the DB.

This is no longer an issue. I double-checked, the image name stays the same, even after submitting the form multiple times. The image path stays the same in the DB.

Closing this since it seems it's been fixed in the meantime.
Again, sorry for the frustrating experience, and for not seeing this sooner.

Cheers!
